### PR TITLE
Improve method filterPath + other enhancements

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | -

Improve method `filterPath`:

- Add option "filter"
- Add support string value for options "only" and "except"
- Set option "basePath" is optional

Other:

- Add functions `is_array` and `is_string` to section `use`
- Add more tests
- Replace phan to psalm
- Add infection
- Improve testing
- Improve readme

### CODE REVIEW

File: src/FileHelper.php [hunk 1]
```diff
@@ -464,15 +467,33 @@ public static function filterPath(string $path, array $options): bool
...
+        if (isset($options['filter'])) {
+            $result = call_user_func($options['filter'], $path);
+            if (is_bool($result)) {
+                return $result;
+            }
+        }
+
         if (!empty($options['except'])) {
-            if ((self::lastExcludeMatchingFromList($options['basePath'], $path, $options['except'])) !== null) {
+            if (
```
Comment (rustamwin) [inline src/FileHelper.php:478]: Can be merged parent one?

File: src/FileHelper.php [hunk 2]
```diff
@@ -464,15 +467,33 @@ public static function filterPath(string $path, array $options): bool
     {
         $path = str_replace('\\', '/', $path);
 
+        if (isset($options['filter'])) {
```
Comment (rustamwin) [inline src/FileHelper.php:470]: What happens if the filter is not callable?

File: src/FileHelper.php (hunk 1 above)
Comment (vjik) [reply src/FileHelper.php:478]: Can, but I think in current variant more understandable. What do you think?

File: src/FileHelper.php (hunk 2 above)
Comment (vjik) [reply src/FileHelper.php:470]: Added exception.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**